### ISSUE TITLE
SEN12MSCR测试问题

### ISSUE THREAD

**Issue by DDKdakuan** [Ly403/EMRDM#14]
你好，非常感谢您开源您的代码和权重，我基于您开源的SEN12MSCR的相关内容进行了测试，测试的文件如下：model:
  base_learning_rate: 1.0e-4
  target: sgm.models.diffusion.ResidualDiffusionEngine
  params:
    input_key: "target"
    mean_key: "S2"
    compile_model: False
    use_ema: True
    use_flash_attn2: False
    ckpt_path: "/opt/DDDDK/cloudremove/EMRDM-main/SEN12MSCR_checkpoint/last.ckpt" 

    sigma_st_config:
      target: sgm.modules.diffusionmodules.sigma2st.EDMSigma2St
      params:
        alpha: 3.0

    denoiser_config:
      target: sgm.modules.diffusionmodules.denoiser.ResidualDenoiser
      params:
        scaling_config:
          target: sgm.modules.diffusionmodules.denoiser_scaling.ResidualEDMScaling
          params:
            sigma_input: 1.0
            sigma_mu: 1.0

    network_wrapper: "sgm.modules.diffusionmodules.wrappers.CloudRemovalWrapper"

    network_config:
      target: sgm.modules.diffusionmodules.k_diffusion.image_transformer.ImageTransformerDenoiserModelInterface
      params:
        in_channels:  28
        out_channels: 13
        patch_size: [1,1]
        widths: [128,256,384,768]
        depths: [2,2,2,2]
        d_ffs: [256,512,768,1536]
        self_attns: [
            {"type": "neighborhood", "d_head": 64, "kernel_size": 7},
            {"type": "neighborhood", "d_head": 64, "kernel_size": 7},
            {"type": "global", "d_head": 64},
            {"type": "global", "d_head": 64},
        ]
        dropout_rate: [0.0,0.0,0.0,0.1]
        mapping_depth: 2
        mapping_width: 768
        mapping_d_ff: 1536
        mapping_dropout_rate: 0.1

    conditioner_config:
      target: sgm.modules.GeneralConditioner
      params:
        emb_models:
          - is_trainable: True
            input_key:  "S1S2"
            ucg_rate: 0.0
            target: sgm.modules.encoders.modules.IndentityEmbedder 

    first_stage_config:
      target: sgm.models.autoencoder.IdentityFirstStage

    loss_fn_config:
      target: sgm.modules.diffusionmodules.loss.ResidualDiffusionLoss
      params:
        loss_weighting_config:
          target: sgm.modules.diffusionmodules.loss_weighting.ResidualEDMWeighting # ResidualSoftMinSnrWeighting
          params:
            sigma_input: 1.0
            sigma_mu: 1.0
        sigma_sampler_config:
          target: sgm.modules.diffusionmodules.sigma_sampling.EDMSampling
          params:
            p_mean: -1.2
            p_std: 1.2

    sampler_config:
      target: sgm.modules.diffusionmodules.sampling.ResidualEulerEDMSampler
      params:
        num_steps: 5
        s_churn: 5.0
        s_tmin: 0.0
        s_tmax: 100000000.0
        s_noise: 1.023

        discretization_config:
          target: sgm.modules.diffusionmodules.discretizer.EDMDiscretization
          params:
            sigma_min: 0.001
            sigma_max: 100.0
    
    to_rgb_config:
      target: sgm.util.S1andS2_to_rgb

data:
  target: sgm.data.base.DataModuleFromConfig
  params:
    batch_size: 1
    num_workers: 8
    wrap: True
    train:
      target: sgm.data.sentinel.sentinel.SEN12MSCRInterface
      params:
        root: "/opt/DDDDK/Dataset/SEN12MS-CR"
        split: "train"
        region: "all"
        rescale: True
        cloud_masks:

    validation:
      target: sgm.data.sentinel.sentinel.SEN12MSCRInterface
      params:
        root: "/opt/DDDDK/Dataset/SEN12MS-CR"
        split: "val"
        region: "all"
        rescale: True
        cloud_masks:

    test:
      target: sgm.data.sentinel.sentinel.SEN12MSCRInterface
      params:
        root: "/opt/DDDDK/Dataset/SEN12MS-CR"
        split: "test"
        region: "all"
        rescale: True
        cloud_masks:

    predict:
      target: sgm.data.sentinel.sentinel.SEN12MSCRInterface
      params:
        root: "/opt/DDDDK/Dataset/SEN12MS-CR"
        split: "test"
        region: "all"
        rescale: True
        cloud_masks:

lightning:
  modelcheckpoint:
    params:
      every_n_train_steps: 5000
      monitor: "RMSE"

  callbacks:
    metrics_over_trainsteps_checkpoint:
      params:
        every_n_train_steps: 25000

    image_logger:
      target: main.ImageLogger
      params:
        disabled: False
        batch_frequency: 2000
        max_images: 64
        increase_log_steps: False
        log_first_step: True
        log_images_kwargs:
          use_ema_scope: False
          N: 64
          n_rows: 16
          return_intermediate: True
          return_denoised: True
          return_add_mu: True
          return_add_noise: True
          return_cond: True

  trainer:
    devices: 1,
    num_sanity_val_steps: 2
    benchmark: True
    accumulate_grad_batches: 1
    max_epochs: 50

测试的命令为：CUDA_VISIBLE_DEVICES=4,5,6,7 python main.py --base configs/example_training/sentinel.yaml --enable_tf32 -t false

中间过程的图像为：mu_noise_shifting_gs-000000_e-000000_b-004000

<img width="1292" height="260" alt="Image" src="https://github.com/user-attachments/assets/ee8cf292-e1ed-41db-8051-97d4b91dfc5c" />

mu_shifting_gs-000000_e-000000_b-004000
<img width="1292" height="260" alt="Image" src="https://github.com/user-attachments/assets/19d9886b-e550-4873-96ee-c7f6c8e4f89e" />
看起来效果也还行，但是最终输出的结果是这样子的，您觉得问题可能出现在哪里，代码我没有任何改动：

────────────────────
       Test metric             DataLoader 0
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
           MAE              0.1754174530506134
          PSNR               13.70605182647705
          RMSE              0.2184729427099228
           SAM               31.56200408935547
          SSIM              0.05794397369027138
       final_LPIPS                  nan
        final_MAE           0.1754172444343567
   final_MAE_cloudfree              nan
    final_MAE_cloudy                nan
       final_PSNR           13.706077575683594
       final_RMSE           0.2184728980064392
  final_RMSE_cloudfree              nan
    final_RMSE_cloudy               nan
        final_SAM            31.56198501586914
       final_SSIM           0.05794403329491615
       final_error                  nan
      final_mean ae                 nan
      final_mean se                 nan
     final_mean var                 nan
         raw_MAE            0.11673440039157867
        raw_PSNR            18.143095016479492
        raw_RMSE            0.15762585401535034
         raw_SAM            13.370469093322754
        raw_SSIM            0.6590080261230469

**Comment by DDKdakuan** [comment]
附一个denoised_gs-000000_e-000000_b-004000的过程图：

<img width="1292" height="260" alt="Image" src="https://github.com/user-attachments/assets/e7962c3c-b739-4a1d-83a0-1ab58388990b" />

**Comment by Ly403** [comment]
您好，首先十分感谢您对我们的工作感兴趣。

从您的描述和对应的图像来看，好像模型没成功加载权重。mu_noise_shifting_表示的是加噪过程中的图像，mu_shifting_表示加噪过程中不加噪声的图像（即只加mu）。denoised是模型的输出，intermediate是每一步的带噪声图像x_t。所以从denoised的结果看，模型几乎没有去噪能力。我又尝试在我本地尝试了一下测试，发现没什么问题。您可以看看加载权重的时候是否正确，一般来说加载权重后控制台会输出这样一句话：“Restored from /remote-home/liuyi/EMRDM-main/last.ckpt with 0 missing and 0 unexpected keys”。您可以把程序输出重定向到文本文件看看这句话有没有输出，或者有没有missing或者unexpected的keys。

希望这些能帮助到您，若您还是没有解决问题，可以继续和我联系，

**Comment by DDKdakuan** [comment]
感谢作者的回复！问题已经得到解决，谢谢！

**Comment by duangtg** [comment]
请问一下 这个问题是怎么解决的呢？我使用了作者放在Google云盘上的权重，然后没有修改相关的代码；最终输出的结果似乎也没有成功加载网络；但是命令台已经输出了
“LOGDIR: logs/2026-07-03T07-31-20_example_training-sentinel
Running on GPUs 0,
Restored from checkpoints/last.ckpt with 0 missing and 0 unexpected keys
Model has 39.1335M parameters.
GPU available: True (cuda), used: True”
GT和输出结果分别是

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/de4796cd-5203-47b0-9835-e681f80f3d33" />

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/e79a9bcf-f7ed-4034-8d89-47b66a4b2bfe" />

**Comment by Ly403** [comment]
@duangtg 您好，能看一下您的yaml文件吗，您是通过什么方式加载权重进去的，看起来好像用的是resume是吗？如果要是推理的话应该用 model.params.ckpt_path，resume是提供给断点重训的。不知道这能不能帮助到您。

**Comment by duangtg** [comment]
感谢您的回复，我仅修改了model.params.ckpt_path以及对应的文件路径

**Comment by duangtg** [comment]
model:
  base_learning_rate: 1.0e-4
  target: sgm.models.diffusion.ResidualDiffusionEngine
  params:
    input_key: "target"
    mean_key: "S2"
    compile_model: False
    use_ema: True
    use_flash_attn2: False
    ckpt_path: "./checkpoints/last.ckpt" # your checkpoint path

    sigma_st_config:
      target: sgm.modules.diffusionmodules.sigma2st.EDMSigma2St
      params:
        alpha: 3.0

    denoiser_config:
      target: sgm.modules.diffusionmodules.denoiser.ResidualDenoiser
      params:
        scaling_config:
          target: sgm.modules.diffusionmodules.denoiser_scaling.ResidualEDMScaling
          params:
            sigma_input: 1.0
            sigma_mu: 1.0

    network_wrapper: "sgm.modules.diffusionmodules.wrappers.CloudRemovalWrapper"

    network_config:
      target: sgm.modules.diffusionmodules.k_diffusion.image_transformer.ImageTransformerDenoiserModelInterface
      params:
        in_channels:  28
        out_channels: 13
        patch_size: [1,1]
        widths: [128,256,384,768]
        depths: [2,2,2,2]
        d_ffs: [256,512,768,1536]
        self_attns: [
            {"type": "neighborhood", "d_head": 64, "kernel_size": 7},
            {"type": "neighborhood", "d_head": 64, "kernel_size": 7},
            {"type": "global", "d_head": 64},
            {"type": "global", "d_head": 64},
        ]
        dropout_rate: [0.0,0.0,0.0,0.1]
        mapping_depth: 2
        mapping_width: 768
        mapping_d_ff: 1536
        mapping_dropout_rate: 0.1

    conditioner_config:
      target: sgm.modules.GeneralConditioner
      params:
        emb_models:
          - is_trainable: True
            input_key:  "S1S2"
            ucg_rate: 0.0
            target: sgm.modules.encoders.modules.IndentityEmbedder
          # - is_trainable: True
          #   input_key:  "S1"
          #   ucg_rate: 0.0
          #   target: sgm.modules.encoders.modules.ImageTransformerEncoderInterface
          #   params:
          #     in_channels: 2
          #     patch_size: [4,4]
          #     widths: [256,512,768]
          #     depths: [2,2,16]
          #     d_ffs: [512,1024,1536]
          #     self_attns: [
          #         {"type": "neighborhood", "d_head": 64, "kernel_size": 7},
          #         {"type": "neighborhood", "d_head": 64, "kernel_size": 7},
          #         {"type": "global", "d_head": 64},
          #     ]
          #     dropout_rate: [0.0,0.0,0.0]

    first_stage_config:
      target: sgm.models.autoencoder.IdentityFirstStage

    loss_fn_config:
      target: sgm.modules.diffusionmodules.loss.ResidualDiffusionLoss
      params:
        loss_weighting_config:
          target: sgm.modules.diffusionmodules.loss_weighting.ResidualEDMWeighting # ResidualSoftMinSnrWeighting
          params:
            sigma_input: 1.0
            sigma_mu: 1.0
        sigma_sampler_config:
          target: sgm.modules.diffusionmodules.sigma_sampling.EDMSampling
          params:
            p_mean: -1.2
            p_std: 1.2

    sampler_config:
      target: sgm.modules.diffusionmodules.sampling.ResidualEulerEDMSampler
      params:
        num_steps: 5
        s_churn: 5.0
        s_tmin: 0.0
        s_tmax: 100000000.0
        s_noise: 1.023

        discretization_config:
          target: sgm.modules.diffusionmodules.discretizer.EDMDiscretization
          params:
            sigma_min: 0.001
            sigma_max: 100.0

    to_rgb_config:
      target: sgm.util.S1andS2_to_rgb

data:
  target: sgm.data.base.DataModuleFromConfig
  params:
    batch_size: 1
    num_workers: 8
    wrap: True
    train:
      target: sgm.data.sentinel.sentinel.SEN12MSCRInterface
      params:
        root: "../SEN12MSCR"
        split: "train"
        region: "all"
        rescale: True
        cloud_masks:

    validation:
      target: sgm.data.sentinel.sentinel.SEN12MSCRInterface
      params:
        root: "../SEN12MSCR"
        split: "val"
        region: "all"
        rescale: True
        cloud_masks:

    test:
      target: sgm.data.sentinel.sentinel.SEN12MSCRInterface
      params:
        root: "../SEN12MSCR"
        split: "test"
        region: "all"
        rescale: True
        cloud_masks:

    predict:
      target: sgm.data.sentinel.sentinel.SEN12MSCRInterface
      params:
        root: "../SEN12MSCR"
        split: "test"
        region: "all"
        rescale: True
        cloud_masks:

lightning:
  modelcheckpoint:
    params:
      every_n_train_steps: 5000
      monitor: "RMSE"

  callbacks:
    metrics_over_trainsteps_checkpoint:
      params:
        every_n_train_steps: 25000

    image_logger:
      target: main.ImageLogger
      params:
        disabled: False
        batch_frequency: 2000
        max_images: 64
        increase_log_steps: False
        log_first_step: True
        log_images_kwargs:
          use_ema_scope: False
          N: 64
          n_rows: 16
          return_intermediate: True
          return_denoised: True
          return_add_mu: True
          return_add_noise: True
          return_cond: True

  trainer:
    devices: 0,
    num_sanity_val_steps: 2
    benchmark: True
    accumulate_grad_batches: 1
    max_epochs: 100

**Comment by duangtg** [comment]
感谢作者的回复！问题已经得到解决，谢谢！
